src/github/pullRequestModel.ts

```diff
@@ -1751,42 +1752,58 @@ export class PullRequestModel extends IssueModel<PullRequest> implements IPullRe
 		}
 	}
 
-	async markFileAsViewed(filePathOrSubpath: string, event: boolean): Promise<void> {
-		const { mutate, schema } = await this.githubRepository.ensure();
-		const fileName = filePathOrSubpath.startsWith(this.githubRepository.rootUri.path) ?
-			filePathOrSubpath.substring(this.githubRepository.rootUri.path.length + 1) : filePathOrSubpath;
-		await mutate<void>({
-			mutation: schema.MarkFileAsViewed,
-			variables: {
-				input: {
-					path: fileName,
-					pullRequestId: this.graphNodeId,
-				},
-			},
-		});
-
-		this.setFileViewedState(fileName, ViewedState.VIEWED, event);
-	}
+	private readonly BATCH_SIZE = 100;
 
-	async unmarkFileAsViewed(filePathOrSubpath: string, event: boolean): Promise<void> {
-		const { mutate, schema } = await this.githubRepository.ensure();
-		const fileName = filePathOrSubpath.startsWith(this.githubRepository.rootUri.path) ?
-			filePathOrSubpath.substring(this.githubRepository.rootUri.path.length + 1) : filePathOrSubpath;
-		await mutate<void>({
-			mutation: schema.UnmarkFileAsViewed,
-			variables: {
-				input: {
-					path: fileName,
-					pullRequestId: this.graphNodeId,
-				},
-			},
-		});
+	async markFiles(filePathOrSubpaths: string[], event: boolean, state: 'viewed' | 'unviewed'): Promise<void> {
+		const { mutate } = await this.githubRepository.ensure();
+		const pullRequestId = this.graphNodeId;
 
-		this.setFileViewedState(fileName, ViewedState.UNVIEWED, event);
-	}
+		const allFilenames = filePathOrSubpaths
+			.map((f) =>
+				f.startsWith(this.githubRepository.rootUri.path)
+					? f.substring(this.githubRepository.rootUri.path.length + 1)
+					: f
+			);
 
-	async unmarkAllFilesAsViewed(): Promise<void[]> {
-		return Promise.all(Array.from(this.fileChanges.keys()).map(change => this.unmarkFileAsViewed(change, true)));
+		const mutationName = state === 'viewed'
+			? 'markFileAsViewed'
+			: 'unmarkFileAsViewed';
+
+		// We only ever send 100 mutations at once. Any more than this and
+		// we risk a timeout from GitHub.
+		for (let i = 0; i < allFilenames.length; i += this.BATCH_SIZE) {
+			const batch = allFilenames.slice(i, i + this.BATCH_SIZE);
+			// See below for an example of what a mutation produced by this
+			// will look like
+			const mutation = gql`mutation Batch${mutationName}{
```

Comment:
Related comment: https://github.com/microsoft/vscode-pull-request-github/pull/4700#discussion_r1165606886

Unfortunately I couldn't find a way around this inlining. It seems when we're creating a query dynamically, this is how it needs to be. Very happy to be shown a different way though; inlining doesn't feel great.

Reply:
The benefit we get from batching is worth the inlining. Thanks for checking